Update plugins_loaded priority algorithm.

zinigor opened the pull request:
Fixes #14757 

This is an exact copy of the PR: #14771 I have copied it verbatim to allow Jetpack team to add our own commits. Thanks a lot for a great idea @strategio !

#### Changes proposed in this Pull Request:
Instead of triggering an error if some other plugin has a higher top priority than us, we'll reposition our callback to be the first one.

#### Testing instructions:
**I did not test this change on the application level and I am not familiar enough with this plugin to provide test steps.**

#### Proposed changelog entry for your changes:
Prevent a PHP notice when a 3rd part plugin is loaded before Jetpack.

strategio committed:
Reposition our configuration action if it's not the first priority

https://github.com/Automattic/jetpack/issues/14757

strategio committed:
FIXUP Use another approach to remove and re-add actions on the highest priority

strategio committed:
FIXUP Recover CSS file

zinigor committed:
Added a check to avoid re-setting actions without need.

zinigor committed:
Attempting to reduce 'cognitive complexity'.

strategio reviewed (dismissed):
Looks good to me @zinigor, thanks for adjusting the code :)

kbrown9 commented on class.jetpack.php:
```diff
@@ -378,6 +378,14 @@ class Jetpack {
+	 *
+	 * @var Integer the priority value.
+	 */
+	protected $highest_priority = null;
```
Just a nitpick, but what do you think of using a different variable name here? We also have the local `$highest_priority` variable in the `add_configure_hook()` method. The two variables have the same name but represent different things.

Maybe something like `$configure_hook_priority`?

zinigor replied:
You're right, thanks, I'll change that!

kbrown9 commented on class.jetpack.php:
```diff
@@ -568,28 +576,36 @@ static function delete_active_modules() {
-			remove_action( 'plugins_loaded', array( $this, 'configure' ), $current_priority );
+		// If there is no handler set for the plugins_loaded hook yet, set it and return.
+		if ( ! isset( $wp_filter['plugins_loaded'] ) ) {
+			add_action( 'plugins_loaded', array( $this, 'configure' ), 0 );
```
Should we set the `$this->highest_priority` value to 0 here before returning?

strategio replied:
You are right @kbrown9. I would even go a step further by setting it as the default value of the private property. And this would also address your [second comment](https://github.com/Automattic/jetpack/pull/14807/files#r385211681).

zinigor replied:
Fixed that, thank you!

kbrown9 commented on class.jetpack.php:
```diff
@@ -568,28 +576,36 @@ static function delete_active_modules() {
+
+		foreach ( $actions_on_highest_priority as $action ) {
+			add_action( 'plugins_loaded', $action['function'], $highest_priority, $action['accepted_args'] );
+		}
```
If `configure` is already hooked at some priority and needs to be moved to a higher priority, we need to remove the original hook, right? I don’t think that’s being done.

zinigor replied:
Yes, you're right, great catch! I have fixed that and added a test against that.

kbrown9 commented:
This is a clever approach! I agree that testing this is difficult. What do you think of adding some unit tests for the `add_configure_hook` method? I think that will help increase our confidence in the correctness of the method.

zinigor committed:
Added tests for add_configure_hook.

kbrown9 commented on tests/php/general/test_class.jetpack.php:
```diff
@@ -1129,6 +1129,98 @@ public function test_wp_getOptions_hook_in_place() {
+	 *
+	 * @covers Jetpack::add_configure_hook
+	 */
+	public function test_add_configure_hook_adds_default_hook() {
```
It looks like a lot of hooks are added to plugins_loaded before the test is executed, so this doesn’t actually test the default priority hook. `test_add_configure_hook_saves_priority_on_no_hooks_present()` does test the default priority hook because all of the plugins_loaded hooks are removed before calling Jetpack::init().

kbrown9 commented on class.jetpack.php:
```diff
@@ -577,23 +585,38 @@ static function delete_active_modules() {
+			// Priority is in reverse order, so less is higher.
+			&& $this->configure_hook_priority <= $highest_priority
+		) {
+			return;
```
A unit test failed because `Jetpack::configure_hook_priority` had a value but the configure hook on plugins_loaded didn’t actually exist. I’m not sure if it’s possible to get into this state in production. To be on the safe side, what do you think of checking that the hook exists and adding it if it doesn’t before returning here?

zinigor replied:
Done in the latest commit, thanks!

kbrown9 reviewed (commented):
I looked into the failing unit test:

> 1) WP_Test_Jetpack::test_add_configure_hook_adds_default_hook
> Failed asserting that false is identical to 0.

I think I understand what’s happening. It looks like PHPUnit resets the globals before each test, so `$wp_filters` is reset. However, the Jetpack singleton object (and the value of `Jetpack::configure_hook_priority`) persists over all of the unit tests. 

Here are the details:

The first configure hook test, `test_add_configure_hook_saves_priority_on_no_hooks_present()`, sets `Jetpack::configure_hook_priority` to 0.

In the second configure_hook test, `test_add_configure_hook_adds_default_hook()`, `Jetpack::configure_hook_priority` still equals 0. This conditional in `Jetpack::configure_hook()` evaluates to true:

```
// If our hook is set, and it's still at highest priority, we don't need to do anything.
		if (
			isset( $this->configure_hook_priority )

			// Priority is in reverse order, so less is higher.
			&& $this->configure_hook_priority <= $highest_priority
		) {
			return;
		}
```

As shown above, `Jetpack::configure_hook()` just returns without setting the configure hook.

Then in the `test_add_configure_hook_adds_default_hook()` unit test, `$current_priority` is set to false. Since 0 is expected, the test fails.

```
$current_priority = has_filter( 'plugins_loaded', array( $jetpack, 'configure' ) );
$this->assertSame( 0, $current_priority );
```

To solve this, maybe we should set `$jetpack->configure_hook_priority` to null at the beginning of each of the configure hook unit tests?

zinigor committed:
Added test instance cleanup, added zero defaults.

zinigor commented:
Yeah, not running the whole suite bit me again. Thanks for the catch. I added a reset code to the teardown method, so now every test runs on a fresh Jetpack instance. I think it will serve us better in the end. Running tests on a pre-existing instance can be misleading.

Unfortunately, that also fixed that edge case that you mentioned - having the priority set, but no actual hook. I actually wanted to ask you about that: do you think we need to be fixing that case? It can happen when a plugin deliberately reset all hooks set on `plugins_loaded` for whatever reason. Should we be adding it back?

kbrown9 commented:
> I added a reset code to the teardown method, so now every test runs on a fresh Jetpack instance. I think it will serve us better in the end. Running tests on a pre-existing instance can be misleading.

Good point - that's a better solution.

> Unfortunately, that also fixed that edge case that you mentioned - having the priority set, but no actual hook. I actually wanted to ask you about that: do you think we need to be fixing that case? It can happen when a plugin deliberately reset all hooks set on `plugins_loaded` for whatever reason. Should we be adding it back?

I think we probably should. `Jetpack::add_configure_hook()`’s job is to add the configure hook, and right now it’s using `Jetpack::configure_hook_priority` as a signal that the hook has already been set. But, as you said, a plugin could reset all the hooks on 'plugins_loaded'. If that happens, the signal isn’t correct and `Jetpack::add_configure_hook()` doesn’t do its job.

What do you think?

zinigor committed:
Experiment with adding globals backup to one test suite.

zinigor committed:
Added logic to add configure hook even if the property is set.

zinigor commented:
Thanks for the review, Kim! I have added new logic that makes the configure hook gets added even if there's some value in the property already, but no hook set.

zinigor commented:
Based on the work that happens in the `feature/update_autoloader` branch, this PR may become obsolete in favor of a simpler solution not based on `plugins_loaded`.

jeherve commented:
I'll close this PR for now because of the lack of activity on this. We can always reopen in the future if needed, but it will need a rebase, so it may be easier to start a new PR at this point.
